### PR TITLE
executor,localstore: use default value if column not found in row.

### PR DESCRIPTION
**Pull request by coocood**
Fix failing to update while adding not null column in local store.

TiKV needs to implement this too.

### CODE REVIEW

**Comment by coocood** [comment]
@shenli @zimulala @BusyJay PTAL

File: ddl/ddl_db_test.go [hunk 1]
```diff
@@ -1004,3 +1004,28 @@ func (s *testDBSuite) testRenameTable(c *C, storeStr, sql string) {
...
+	s.tk = testkit.NewTestKit(c, s.store)
+	s.tk.MustExec("use test_db")
+	// for different databases
+	s.tk.MustExec("create table tnn (c1 int primary key auto_increment, c2 int)")
+	s.tk.MustExec("insert tnn (c2) values (0)" + strings.Repeat(",(0)", 99))
+	done := make(chan error, 1)
+	sessionExecInGoroutine(c, s.store, "alter table tnn add column c3 int not null default 3", done)
+	updateCnt := 0
+out:
+	for {
```
Comment (hanfei1991) [inline ddl/ddl_db_test.go]: Why put this into a for loop ?

Comment (coocood) [inline ddl/ddl_db_test.go]: We need to make sure this executes in every schema state.

File: store/localstore/local_region.go [hunk 2]
```diff
@@ -533,12 +533,17 @@ func (rs *localRegion) handleRowData(ctx *selectContext, handle int64, value []b
...
 			_, ok := values[col.GetColumnId()]
-			if !ok {
-				if mysql.HasNotNullFlag(uint(col.GetFlag())) {
-					return false, errors.New("Miss column")
-				}
-				values[col.GetColumnId()] = []byte{codec.NilFlag}
+			if ok {
+				continue
+			}
+			if len(col.DefaultVal) > 0 {
```
Comment (BusyJay) [inline store/localstore/local_region.go]: What if the default value is an empty string?

Comment (coocood) [inline store/localstore/local_region.go]: It is encoded bytes, empty string has a type byte.

File: executor/distsql.go [hunk 3]
```diff
@@ -819,7 +833,20 @@ func (e *XSelectTableExec) doRequest() error {
...
+	selReq.TableInfo.Columns = distsql.ColumnsToProto(e.Columns, e.tableInfo.PKIsHandle)
+	for i, c := range e.Columns {
+		if c.State == model.StatePublic {
+			continue
+		}
+		d, _, err := table.GetColDefaultValue(e.ctx, c)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		selReq.TableInfo.Columns[i].DefaultVal, err = tablecodec.EncodeValue(d)
```
Comment (BusyJay) [inline executor/distsql.go]: I'm not familiar with this, but when the default value of a column is changed twice, which value will be used?

Comment (coocood) [inline executor/distsql.go]: It used what TiDB server passed to coprocessor.
Currently, TiDB only pass default value when the column is not in public state.

File: ddl/ddl_db_test.go [hunk 4]
```diff
@@ -1004,3 +1004,28 @@ func (s *testDBSuite) testRenameTable(c *C, storeStr, sql string) {
...
+
+func (s *testDBSuite) TestAddNotNullColumn(c *C) {
+	defer testleak.AfterTest(c)
+	s.tk = testkit.NewTestKit(c, s.store)
+	s.tk.MustExec("use test_db")
+	// for different databases
+	s.tk.MustExec("create table tnn (c1 int primary key auto_increment, c2 int)")
+	s.tk.MustExec("insert tnn (c2) values (0)" + strings.Repeat(",(0)", 99))
+	done := make(chan error, 1)
+	sessionExecInGoroutine(c, s.store, "alter table tnn add column c3 int not null default 3", done)
```
Comment (zimulala) [inline ddl/ddl_db_test.go]: What about the behavior of executing `alter table tnn add column c3 int not null`?

Comment (zimulala) [inline ddl/ddl_db_test.go]: MySQL fills it with `ZeroValue` when sql_mode is equal to `STRICT_TRANS_TABLES`.

File: table/column.go [hunk 5]
```diff
@@ -241,37 +241,18 @@ func CheckNotNull(cols []*Column, row []types.Datum) error {
 
 // GetColDefaultValue gets default value of the column.
 func GetColDefaultValue(ctx context.Context, col *model.ColumnInfo) (types.Datum, bool, error) {
-	// Check no default value flag.
-	if mysql.HasNoDefaultValueFlag(col.Flag) && col.Tp != mysql.TypeEnum {
```
Comment (zimulala) [inline table/column.go]: Why remove the check of `mysql.HasNoDefaultValueFlag(col.Flag) is true` ?

**Comment by zimulala** [comment]
LGTM

**Pull request merged**